src/blockchain/ain/jellyfish/jellyfish.service.ts

```diff
@@ -183,20 +223,145 @@ export class JellyfishService {
+      vaultFee,
+    );
+  }
+
+  private async generateRawVaultTx(
+    from: string,
+    vault: string,
+    token: number,
+    amount: BigNumber,
+    createVout: (vault: string, script: Script, token: number, amount: BigNumber) => Vout,
+  ): Promise<RawTxDto> {
+    const [fromScript, fromPubKeyHash] = RawTxUtil.parseAddress(from);
+
+    const utxo = await this.utxoProvider.provideForDefiTx(from);
+    return this.generateRawDefiTx(fromScript, fromPubKeyHash, utxo, createVout(vault, fromScript, token, amount));
+  }
+
+  private async generateRawTxForAddPoolLiquidity(
+    from: string,
+    tokenA: number,
+    amountA: BigNumber,
+    tokenB: number,
+    amountB: BigNumber,
+  ): Promise<RawTxDto> {
+    const [fromScript, fromPubKeyHash] = RawTxUtil.parseAddress(from);
+
+    const tokenBalanceA: TokenBalanceUInt32 = { token: tokenA, amount: amountA };
+    const tokenBalanceB: TokenBalanceUInt32 = { token: tokenB, amount: amountB };
+
+    const utxo = await this.utxoProvider.provideForDefiTx(from);
+    return this.generateRawDefiTx(
+      fromScript,
+      fromPubKeyHash,
+      utxo,
+      RawTxUtil.createVoutAddPoolLiquidity(fromScript, [tokenBalanceA, tokenBalanceB]),
+    );
+  }
+
+  private async generateRawTxForRemovePoolLiquidity(from: string, token: number, amount: BigNumber): Promise<RawTxDto> {
+    const [fromScript, fromPubKeyHash] = RawTxUtil.parseAddress(from);
+
+    const utxo = await this.utxoProvider.provideForDefiTx(from);
+    return this.generateRawDefiTx(
+      fromScript,
+      fromPubKeyHash,
+      utxo,
+      RawTxUtil.createVoutRemovePoolLiquidity(fromScript, token, amount),
+    );
+  }
+
+  private async generateRawTxForCompositeSwap(
+    from: string,
+    fromToken: number,
+    fromAmount: BigNumber,
+    toToken: number,
+  ): Promise<RawTxDto> {
+    const [fromScript, fromPubKeyHash] = RawTxUtil.parseAddress(from);
+
+    const utxo = await this.utxoProvider.provideForDefiTx(from);
+    return this.generateRawDefiTx(
+      fromScript,
+      fromPubKeyHash,
+      utxo,
+      RawTxUtil.createVoutCompositeSwap(
+        fromScript,
+        fromToken,
+        fromAmount,
+        toToken,
+        new BigNumber(Config.blockchain.default.maxPrice),
+      ),
+    );
+  }
+
   // --- UTXO MANAGEMENT --- //
   private async executeUnlockUtxos(prevouts: Prevout[], scriptHex: string): Promise<void> {
-    this.utxoProvider.unlockSpentBasedOn(
-      prevouts,
-      RawTxUtil.parseAddressFromScriptHex(scriptHex, JellyfishService.getNetwork()),
-    );
+    this.utxoProvider.unlockSpentBasedOn(prevouts, RawTxUtil.parseAddressFromScriptHex(scriptHex));
   }
 
   // --- HELPER METHODS --- //
-  private createTxAndCalcFee(utxo: UtxoInformation, vins: Vin[], vouts: Vout[], witnesses: Witness[]): RawTxDto {
+  private async generateRawDefiTx(
+    fromScript: Script,
+    fromPubKeyHash: string,
+    utxo: UtxoInformation,
+    vout: Vout,
+    additionalValue = new BigNumber(0),
```

Comment:
Rename `additionalValue` to something with fee? We have a transaction fee and sometimes an operation fee.